File: src/mouse_drawing.cpp
//
// Copyright (c) Manuel Rodriguez.
// Licensed under the MIT license. See LICENSE file in the project root for
// details.
//
// Implementation of MouseDrawing class
#include "mouse_drawing.h"

#include <gdk/gdkkeysyms.h>

#include <iostream>

// MouseDrawing ctor sets the drawing area default width and height
// a surface to store the drawings and enables the handling of mouse
// events.
MouseDrawing::MouseDrawing() {
  set_size_request(drawing_area_w, drawing_area_h);
  this->surface = Cairo::ImageSurface::create(Cairo::Format::FORMAT_ARGB32,
                                              drawing_area_w, drawing_area_h);

  // Enable the events you wish to receive
  add_events(Gdk::BUTTON_PRESS_MASK | Gdk::BUTTON_RELEASE_MASK |
             Gdk::POINTER_MOTION_MASK);
}

MouseDrawing::~MouseDrawing() {}

// Method used to clear the screen, it empties the vector
// used to store the mouse coordinates after a click and
// triggers a redraw while chaning the state to "clear"
void MouseDrawing::clear_screen() {
  this->state = DrawState::clear;
  points.clear();
  queue_draw();
}

// Saves the current surface to a png file called "image"
void MouseDrawing::save_screen() { this->surface->write_to_png("image.png"); }

std::vector<double> MouseDrawing::export_to_vector(int w, int h, double scale) {
  // Create new Cairo surface for the scaled surface
  Cairo::RefPtr<Cairo::ImageSurface> scaled_surface =
      Cairo::ImageSurface::create(Cairo::FORMAT_ARGB32, w, h);
  Cairo::RefPtr<Cairo::Context> scaled_context =
      Cairo::Context::create(scaled_surface);

  double scale_x = (double)w / drawing_area_w;
  double scale_y = (double)h / drawing_area_h;
  scaled_context->scale(scale_x, scale_y);

  // 3. Draw the original surface onto the scaled context of the new surface
  Cairo::RefPtr<Cairo::SurfacePattern> pattern =
      Cairo::SurfacePattern::create(this->surface);
  scaled_context->set_source(pattern);
  scaled_context->paint();

  // 4. Extract pixel data and convert to grayscale vector
  uint8_t *data = scaled_surface->get_data();
  int stride = scaled_surface->get_stride();
  std::vector<double> grayscale_pixels(w * h);

  for (int y = 0; y < h; ++y) {
    for (int x = 0; x < w; ++x) {
      // Cairo typically uses CAIRO_FORMAT_ARGB32, which is BGRX in memory on
      // little-endian systems
      unsigned char *pixel = data + y * stride + x * 4;
      unsigned char blue = pixel[0];
      unsigned char green = pixel[1];
      unsigned char red = pixel[2];

      // Grayscale conversion using luminance formula (approx)
      unsigned char gray = static_cast<unsigned char>(
          0.2126 * red + 0.7152 * green + 0.0722 * blue);
      grayscale_pixels[y * w + x] = gray / scale;
    }
  }

  return grayscale_pixels;
}

// Main function where the drawing is handled
// a state variable is used to decide when to draw and
// when to clear the screen.
// GTK provides and empty cairo context on each on_draw call
// therefore we store the context on a surface and restore the
// context every time we are called.
bool MouseDrawing::on_draw(const Cairo::RefPtr<Cairo::Context> &cr) {
  auto allocation = this->get_allocation();
  // Restore cairo context
  // each time on_draw is called a fresh context is created
  // this way our surface keeps the history
  cr->save();
  cr->set_source(this->surface, 0.0, 0.0);
  cr->restore();
  cr->paint();

  // Check the state to determine our action
  switch (this->state) {
  case DrawState::clear:
    // Set black background
    cr->set_source_rgb(0.0, 0.0, 0.0);
    cr->paint();
    this->state = DrawState::draw;
    break;

  case DrawState::draw:
    cr->save();
    cr->set_line_width(0);
    cr->set_source_rgb(1.0, 1.0, 1.0);
    // Draw circles where the mouse has moved
    for (const auto &point : points) {
      cr->arc(point.x, point.y, brush_size, 0, 2 * M_PI);
      cr->fill();
    }
    cr->restore();
    break;
  default:
    break;
  }

  // Save surface so that we can restore the context on the next call
  Cairo::RefPtr<Cairo::Context> t_context =
      Cairo::Context::create(this->surface);
  t_context->set_source(cr->get_target(), -allocation.get_x(),
                        -allocation.get_y());
  t_context->paint();

  return true;
}

// Checks for left mouse clicks and starts logging the mouse
// coordinates on our points vector to be drawn later
bool MouseDrawing::on_button_press_event(GdkEventButton *event) {
  if (event->button == 1) { // Left mouse button
    this->left_clicked = true;
    this->state = DrawState::draw;
    points.push_back({event->x, event->y});
    std::cout << event->x << " " << event->y << std::endl;
    return true; // Event handled
  }
  return false;
}

// Checks when the click is released to stop drawing.
bool MouseDrawing::on_button_release_event(GdkEventButton *event) {
  if (event->button == 1) {
    this->left_clicked = false;
    return true; // Event handled
  }
  return false;
}

// Logs new coordinates as we move with the button pressed down.
// if movement is detected but the left button was released do nothing.
bool MouseDrawing::on_motion_notify_event(GdkEventMotion *event) {
  if (this->left_clicked) {
    points.push_back({event->x, event->y});
    queue_draw(); // Request a redraw
    std::cout << event->x << " " << event->y << std::endl;
    return true; // Event handled
  }
  return false;
}

File: drawing_window_python/mouse_drawing.py
#!/usr/bin/env python3
"""Mouse drawing widget: provides a drawing area that records mouse strokes,
can clear, save to PNG and export a scaled grayscale vector suitable for
MNIST-like models.
"""
import gi
import cairo
import numpy as np

gi.require_version("Gtk", "3.0")
from gi.repository import Gtk, Gdk  # noqa: E402  (require_version must run before importing Gtk)


class MouseDrawing(Gtk.DrawingArea):
    def __init__(self, width=250, height=250, brush_size=10.0):
        super().__init__()
        self.drawing_area_w = width
        self.drawing_area_h = height
        self.brush_size = brush_size
        self.set_size_request(self.drawing_area_w, self.drawing_area_h)

        # Create surface that stores the drawing history
        self.surface = cairo.ImageSurface(cairo.FORMAT_ARGB32,
                                          self.drawing_area_w,
                                          self.drawing_area_h)

        # State
        self.state = "clear"  # "clear" or "draw"
        self.left_clicked = False
        self.points = []  # list of (x, y)

        # Enable events
        self.add_events(Gdk.EventMask.BUTTON_PRESS_MASK |
                        Gdk.EventMask.BUTTON_RELEASE_MASK |
                        Gdk.EventMask.POINTER_MOTION_MASK)

        # Connect signals
        self.connect("draw", self.on_draw)
        self.connect("button-press-event", self.on_button_press_event)
        self.connect("button-release-event", self.on_button_release_event)
        self.connect("motion-notify-event", self.on_motion_notify_event)

    # Public API
    def clear_screen(self):
        self.state = "clear"
        self.points.clear()
        self.queue_draw()

    def save_screen(self, filename="image.png"):
        self.surface.write_to_png(filename)

    def export_to_vector(self, dtype, w=28, h=28, scale=255.0, target_shape=None):
        """Return a flat NumPy array of grayscale pixels scaled to (w,h).

        dtype: "float32" -> np.float32 values in ~[0.0, 1.0]
               "int8"  -> np.int8 values in [-128..127]
               "uint8" -> np.uint8 values in [0..255]
        """
        # Create scaled surface
        scaled_surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, w, h)
        ctx = cairo.Context(scaled_surface)

        scale_x = float(w) / self.drawing_area_w
        scale_y = float(h) / self.drawing_area_h
        ctx.scale(scale_x, scale_y)

        # Paint original surface into scaled context
        pattern = cairo.SurfacePattern(self.surface)
        ctx.set_source(pattern)
        ctx.paint()

        data = scaled_surface.get_data()
        stride = scaled_surface.get_stride()

        if dtype == "float32":
            out = np.empty(w * h, dtype=np.float32)
        elif dtype == "int8":
            out = np.empty(w * h, dtype=np.int8)
        elif dtype == "uint8":
            out = np.empty(w * h, dtype=np.uint8)
        else:
            # default to uint8
            print(f"Unkwon dtype {dtype}, selecting numpy.uint8")
            out = np.empty(w * h, dtype=np.uint8)

        idx = 0
        for y in range(h):
            row_off = y * stride
            for x in range(w):
                offset = row_off + x * 4
                b = data[offset]
                g = data[offset + 1]
                r = data[offset + 2]
                # Convert to grayscale using luminance
                gray = int(0.2126 * r + 0.7152 * g + 0.0722 * b)
                if dtype == "float":
                    out[idx] = float(gray / scale)
                elif dtype == "int8":
                    out[idx] = np.int8(gray - 128)
                else:  # uint8 or default
                    out[idx] = np.uint8(gray)
                idx += 1

        # If caller requests a specific target_shape (e.g. model input shape)
        # validate product and reshape accordingly. target_shape should be a
        # tuple compatible with NumPy reshape, for example (1,28,28,1).
        if target_shape is not None:
            # ensure it's a tuple
            if not isinstance(target_shape, (tuple, list)):
                raise TypeError("target_shape must be a tuple or list")
            if int(np.prod(target_shape)) != out.size:
                raise ValueError(
                    f"target_shape product {int(np.prod(target_shape))} does not match image size {out.size}"
                )
            return out.reshape(tuple(target_shape))

        return out

    # Event handlers
    def on_draw(self, widget, cr):
        # Paint stored surface
        cr.set_source_surface(self.surface, 0, 0)
        cr.paint()

        if self.state == "clear":
            cr.set_source_rgb(0, 0, 0)
            cr.paint()
            self.state = "draw"

        if self.state == "draw":
            cr.save()
            cr.set_line_width(0)
            cr.set_source_rgb(1, 1, 1)
            for (x, y) in self.points:
                cr.arc(x, y, self.brush_size, 0, 2 * 3.14159)
                cr.fill()
            cr.restore()

        # Save current cr into the surface so next draw will restore it
        # We must copy the widget's target into our stored surface using
        # the widget allocation offset (negative) — this mirrors the C++ code
        # which avoids shrinking/offset issues when GTK provides fresh contexts.
        allocation = self.get_allocation()
        t_ctx = cairo.Context(self.surface)
        # cr.get_target() is the widget surface; offset it so the stored
        # surface receives the correct region
        t_ctx.set_source_surface(cr.get_target(), -allocation.x, -allocation.y)
        t_ctx.paint()

        return False

    def on_button_press_event(self, widget, event):
        if event.button == 1:
            self.left_clicked = True
            self.state = "draw"
            self.points.append((event.x, event.y))
            return True
        return False

    def on_button_release_event(self, widget, event):
        if event.button == 1:
            self.left_clicked = False
            return True
        return False

    def on_motion_notify_event(self, widget, event):
        if self.left_clicked:
            self.points.append((event.x, event.y))
            self.queue_draw()
            return True
        return False
